File: 1/2020.10.06/145.Binary Tree Postorder Traversal/python/solution.py
class Solution:
    def postorderTraversal(self, root: TreeNode) -> List[int]:
        # recursively, one line
        return self.postorderTraversal(root.left) + self.postorderTraversal(root.right) + [root.val] if root else []

    def postorderTraversal(self, root: TreeNode) -> List[int]:
        # iteratively
        if not root:
            return []
        stack, res = [], []
        prev = None
        while stack or root:
            while root:
                stack.append(root)
                root = root.left
            root = stack.pop()
            if not root.right or root.right == prev:
                res.append(root.val)
                prev = root
                root = None
            else:
                stack.append(root)
                root = root.right
        return res

File: 1/2020.10.06/145.Binary Tree Postorder Traversal/java/solution.java
class Solution {
    public List<Integer> postorderTraversal(TreeNode root) {
        // iteratively
        List<Integer> res = new ArrayList<>();
        if (root == null) {
            return res;
        }
        Deque<TreeNode> stack = new LinkedList<>();
        TreeNode prev = null;
        while (!stack.isEmpty() || root != null) {
            while (root != null) {
                stack.offerLast(root);
                root = root.left;
            }
            root = stack.pollLast();
            if (root.right == null || root.right == prev) {
                res.add(root.val);
                prev = root;
                root = null;
            } else {
                stack.offerLast(root);
                root = root.right;
            }
        }
        return res;
    }
}
